dynawo/sources/Modeler/DataInterface/PowSyblIIDM/test/TestRatioTapChangerInterface.cpp

```diff
@@ -0,0 +1,462 @@
+  ASSERT_EQ(Ifce.getNbTap(), 0);             // getNbTap() is the number of steps!
+  ASSERT_EQ(Ifce.getLowPosition(), 1L);      // because tag_YYHY
+  ASSERT_EQ(Ifce.getCurrentPosition(), 2L);  // because tag_UUHU
+
+  Ifce.setCurrentPosition(3L);  // because tag_HHHG
+  ASSERT_EQ(Ifce.getCurrentPosition(), 3L);
+
+  ASSERT_THROW(Ifce.setCurrentPosition(4L), std::exception);
+  ASSERT_THROW(Ifce.setCurrentPosition(-1L), std::exception);
+  ASSERT_THROW(Ifce.setCurrentPosition(0L), std::exception);
+
+  ASSERT_EQ(Ifce.getCurrentPosition(), 3L);
+  ASSERT_THROW(Ifce.getCurrentR(), std::out_of_range);  // 'steps_' must be initialized before accessing values
+
+  for (long i = Ifce.getLowPosition();; i++) {
+    try {
+      Ifce.setCurrentPosition(i);
+    } catch (...) {
+      break;
+    }
+    const auto& x = rTapChanger.getStep(i);
+    powsybl::iidm::RatioTapChangerStep S(x.getRho(), x.getR(), x.getX(), x.getG(), x.getB());
+    Ifce.addStep(boost::shared_ptr<StepInterface>(new StepInterfaceIIDM(S)));
+  }
+  ASSERT_EQ(Ifce.getNbTap(), 3);
+  ASSERT_EQ(Ifce.getNbTap(), rTapChanger.getStepCount());
+
+  Ifce.setCurrentPosition(1L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentRho(), 13.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentR(), 12.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentX(), 14.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentG(), 11.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentB(), 10.0L);
+  Ifce.setCurrentPosition(2L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentRho(), 18.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentR(), 17.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentX(), 19.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentG(), 16.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentB(), 15.0L);
+  Ifce.setCurrentPosition(3L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentRho(), 23.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentR(), 22.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentX(), 24.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentG(), 21.0L);
+  ASSERT_DOUBLE_EQ(Ifce.getCurrentB(), 20.0L);
+
+  ASSERT_TRUE(Ifce.hasLoadTapChangingCapabilities());
+  ASSERT_DOUBLE_EQ(Ifce.getTargetV(), 25.0);
+
+  ASSERT_TRUE(Ifce.getRegulating());
+  ASSERT_EQ(Ifce.getTerminalRefId(), "LOAD1");
+  ASSERT_EQ(Ifce.getTerminalRefSide(), "");  // Always "" for a 2WT --> Arbitrary DG
+  const powsybl::iidm::Terminal& terminal = network.getLoad("LOAD2").getTerminal();
+  rTapChanger.setRegulationTerminal(stdcxx::ref<powsybl::iidm::Terminal>(terminal));
+  ASSERT_EQ(Ifce.getTerminalRefId(), "LOAD2");
+  ASSERT_EQ(Ifce.getTerminalRefSide(), "");
+
+  rTapChanger.setRegulating(false);
+  rTapChanger.setTargetV(stdcxx::nan());
+  ASSERT_FALSE(Ifce.getRegulating());
+  ASSERT_DOUBLE_EQ(Ifce.getTargetV(), 99999.9L);
+  ASSERT_EQ(Ifce.getTerminalRefId(), "");
+  ASSERT_EQ(Ifce.getTerminalRefSide(), "");
+
+  rTapChanger.setTargetV(666.666);
+  rTapChanger.setRegulating(true);
+  ASSERT_TRUE(Ifce.getRegulating());
+  ASSERT_DOUBLE_EQ(Ifce.getTargetV(), 666.666L);
+
+  // Ifce.sanityCheck( Parent) : no direct testing since private
```

Comment:
There should be a check of this with badly formed RatioTapChangers (as it is called in the builder)

Reply:
I agree with you, but I cannot build a bad powsybl RatioTapChanger  in this version. It seems that it has been improved. I have the same trouble with other parts: I cannot set bad objects, so all testing parts of powsybl badly-formed objects seems useless. Anyway I add another case to test, but cannot test since all powsybl objects are good ones.

Reply:
ok, so if you tested that this case cannot happen then we should remove this method

Reply:
I agree; this private method is only called by ctor and only tests if argument passed to the ctor is valid. I couln't build an invalid powsybl object. Only point is that coverage says we did not test this lines. I get rid of the function sanityCheck().

Reply:
done